test/integration/controller_instance_test.go

```diff
@@ -0,0 +1,1475 @@
+		broker:                       getTestBroker(),
+		instance:                     getTestInstance(),
+		skipVerifyingInstanceSuccess: true,
+		preCreateInstance: func(ct *controllerTest) {
+			otherPlanName := "otherplanname"
+			otherPlanID := "other-plan-id"
+			catalogResponse := ct.osbClient.CatalogReaction.(*fakeosb.CatalogReaction).Response
+			catalogResponse.Services[0].PlanUpdatable = truePtr()
+			catalogResponse.Services[0].Plans = []osb.Plan{
+				{
+					Name:        otherPlanName,
+					Free:        truePtr(),
+					ID:          otherPlanID,
+					Description: "another test plan",
+				},
+			}
+
+			ct.broker.Spec.RelistRequests++
+			if _, err := ct.client.ClusterServiceBrokers().Update(ct.broker); err != nil {
+				t.Fatalf("error updating Broker: %v", err)
+			}
+			if err := util.WaitForClusterServicePlanToExist(ct.client, otherPlanID); err != nil {
+				t.Fatalf("error waiting for ClusterServiceClass to exist: %v", err)
+			}
+			if err := util.WaitForClusterServicePlanToNotExist(ct.client, testPlanExternalID); err != nil {
+				t.Fatalf("error waiting for ClusterServiceClass to not exist: %v", err)
+			}
+
+		},
+	}
+	ct.run(func(ct *controllerTest) {
+		condition := v1beta1.ServiceInstanceCondition{
+			Type:   v1beta1.ServiceInstanceConditionReady,
+			Status: v1beta1.ConditionFalse,
+			Reason: "ReferencesNonexistentServicePlan",
+		}
+		if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, condition); err != nil {
+			t.Fatalf("error waiting for instance condition: %v", err)
+		}
+	})
+}
+
+func TestCreateServiceInstanceAsynchronous(t *testing.T) {
+	dashURL := testDashboardURL
+	key := osb.OperationKey(testOperation)
+
+	cases := []struct {
+		name        string
+		osbResponse *osb.ProvisionResponse
+	}{
+		{
+			name: "asynchronous provision with operation key",
+			osbResponse: &osb.ProvisionResponse{
+				Async:        true,
+				DashboardURL: &dashURL,
+				OperationKey: &key,
+			},
+		},
+		{
+			name: "asynchronous provision without operation key",
+			osbResponse: &osb.ProvisionResponse{
+				Async:        true,
+				DashboardURL: &dashURL,
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ct := &controllerTest{
+				t:        t,
+				broker:   getTestBroker(),
+				instance: getTestInstance(),
+				setup: func(ct *controllerTest) {
+					ct.osbClient.ProvisionReaction = &fakeosb.ProvisionReaction{
+						Response: tc.osbResponse,
+					}
+				},
+			}
+			ct.run(func(ct *controllerTest) {
+				condition := v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionReady,
+					Status: v1beta1.ConditionTrue,
+					Reason: "ProvisionedSuccessfully",
+				}
+				if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, condition); err != nil {
+					t.Fatalf("error waiting for instance condition: %v", err)
+				}
+			})
+		})
+	}
+}
+
+func TestDeleteServiceInstance(t *testing.T) {
```

Comment:
For completeness, we should probably do some deletion tests for ServiceInstances that do not have any bindings.

Reply:
Done.